### PR TITLE
Interpolating points

### LINKED ISSUE #2992: Interpolating point clouds
Add a point cloud version that can be used to interpolate between multiple datasets, with different positions

(Similarly to the one by the cosmic life students)

### PR DESCRIPTION
**Pull request by WeirdRubberDuck**
Adds a new type of point cloud, that can be used to interpret the rows in one data file as different time steps of the same set of objects. 

Supports interpolation of positions and data values used for color and size mapping. Positions may also be interpolated using spline interpolation

See example asset. And here is a video demonstrating. I have verified that it works well with a larger number of points, as well as with data files with missing values. 

https://github.com/OpenSpace/OpenSpace/assets/8808894/eb031452-5c91-4ded-92fe-45c552b4b642

closes #2992

### CODE REVIEW

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+        _prevInterpolationValue = _interpolation.value;
+    });
+
+    _interpolation.useSpline.onChange([this]() {
+        _dataIsDirty = true;
+        _shouldReinitializeBufferdata = true;
+    });
+
+    // This property is mostly for show in the UI, but also used to tell how many points
+    // shoudl be rendered. So make sure it is updated once we know the number of
```
Comment (WeirdRubberDuck) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp:300]:
```suggestion
    // should be rendered. So make sure it is updated once we know the number of
```

File: data/assets/examples/pointclouds/interpolated_points.asset [hunk 2]
```diff
@@ -0,0 +1,92 @@
+local Points = {
+  Identifier = "Example_InterpolatedPoints_ColorMapped",
+  Renderable = {
+    Type = "RenderableInterpolatedPoints",
+    -- The dataset here is just a linearly expanding dataset, where the points move in
+    -- a straight line
+    File = asset.resource("data/interpolation_expand.csv"),
+    -- Specify how many objects the rows in the dataset represent. Here, the dataset is
+    -- consists of 10 objects with positions at 6 different time steps. This information
+    -- is reqiired
```
Comment (alexanderbock) [inline data/assets/examples/pointclouds/interpolated_points.asset:10]:
```suggestion
    -- is required
```

File: data/assets/examples/pointclouds/interpolated_points.asset [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    -- Reduce the scale of the points a bit compared to default, so we see them more clearly
+    SizeSettings = {
+      ScaleExponent = 3.0
+    }
+  },
+  GUI = {
+    Name = "Interpolating Points (Spline)",
+    Path = "/Example/Interpolated Point Clouds",
+    Description = [[Example of interpolating points with spline-based interpolation
+      for the position. This leads to smoother transitions at the knots of the
```
Comment (alexanderbock) [inline data/assets/examples/pointclouds/interpolated_points.asset:59]:
```suggestion
      for the position. This leads to smoother transitions at the nodes of the
```
?

File: data/assets/examples/pointclouds/interpolated_points.asset [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      interpolation. Try disabling the spline interpolation in the GUI (under
+      Renderable->Interpolation)to see the difference.]]
```
Comment (alexanderbock) [inline data/assets/examples/pointclouds/interpolated_points.asset:61]:
```suggestion
      Renderable->Interpolation) to see the difference.]]
```

File: data/assets/examples/pointclouds/interpolated_points.asset [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+asset.export(Points)
+asset.export(Points_Smoothed)
+
+
+
+asset.meta = {
+  Name = "Example - Interpolated Point Clouds",
+  Version = "1.0",
+  Description = [[Example of point clouds that support interpolation. One uses a linear
+    motion and color mapping. the other example uses a spline interpolation for the
```
Comment (alexanderbock) [inline data/assets/examples/pointclouds/interpolated_points.asset:85]:
```suggestion
    motion and color mapping. The other example uses a spline interpolation for the
```

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+    addProperty(interpolateToEnd);
+    addProperty(interpolateToStart);
+    addProperty(interpolateToNextStep);
+    addProperty(interpolateToPrevStep);
+    addProperty(speed);
+
+    goToNextStep.onChange([this]() {
+        float prevValue = glm::floor(value);
+        value = glm::min(prevValue + 1.f, value.maxValue());
+        });
```
Comment (alexanderbock) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp:231]:
```suggestion
    });
```

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    addProperty(goToPrevStep);
+
+    nSteps.setReadOnly(true);
+    addProperty(nSteps);
+
+    addProperty(useSpline);
+}
+
+RenderableInterpolatedPoints::RenderableInterpolatedPoints(
+    const ghoul::Dictionary& dictionary)
```
Comment (alexanderbock) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp:248]: Should be right-indented

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+    if (isAtKnot()) {
+        t0 = t0 - 1.f;
+    }
+
+    const float maxTValue = _interpolation.value.maxValue();
+    float maxAllowedT0 = glm::max(maxTValue - 1.f, 0.f);
+    t0 = glm::clamp(t0, 0.f, maxAllowedT0);
+    return t0;
+}
+
```
Comment (alexanderbock) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.cpp:543]:
```suggestion
```

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ ****************************************************************************************/
+
+#ifndef __OPENSPACE_MODULE_BASE___RENDERABLEINTERPOLATEDPOINTS___H__
+#define __OPENSPACE_MODULE_BASE___RENDERABLEINTERPOLATEDPOINTS___H__
+
+#include <modules/base/rendering/pointcloud/renderablepointcloud.h>
+
+#include <openspace/properties/scalar/boolproperty.h>
+#include <openspace/properties/scalar/uintproperty.h>
+#include <openspace/properties/scalar/floatproperty.h>
```
Comment (alexanderbock) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.h:32]:
```suggestion
#include <openspace/properties/scalar/floatproperty.h>
#include <openspace/properties/scalar/uintproperty.h>
```

File: modules/base/rendering/pointcloud/renderableinterpolatedpoints.h [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    std::vector<float> createDataSlice() override;
+
+    void initializeBufferData();
+    void updateBufferData() override;
+
+private:
+    bool isAtKnot() const;
+    float computeCurrentLowerValue() const;
+
+    struct Interpolation : properties::PropertyOwner {
```
Comment (alexanderbock) [inline modules/base/rendering/pointcloud/renderableinterpolatedpoints.h:77]:
```suggestion
    struct Interpolation : public properties::PropertyOwner {
```

File: modules/base/shaders/pointcloud/billboardpoint_interpolated_vs.glsl [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  // check 0 and 1 expicitly
+  if (abs(t) < Epsilon) {
+      return v0;
+  }
+  if (abs(1.0 - t) < Epsilon) {
+      return v1;
+  }
+  bool isMissing = isnan(v0) || isnan(v1);
+  return isMissing ? NaN : mix(v0, v1, t);
+};
```
Comment (alexanderbock) [inline modules/base/shaders/pointcloud/billboardpoint_interpolated_vs.glsl:60]:
```suggestion
}
```

**Review by alexanderbock — APPROVED** [review]

**Pull request merged**